### DISCUSSION TITLE
data_files removal: pip / setuptools is going to be not nice with GTK applications

### DISCUSSION THREAD

**Discussion by jeromerobert** [pdfarranger/pdfarranger#860] [General]
There is a long discussion about that [here](https://discuss.python.org/t/should-there-be-a-new-standard-for-installing-arbitrary-data-files/7853).

Some key points:

* `data_files` is a distutils/setuptools feature we use to install icons, translations, and freedesktop files.
* `wheel` is a standard for python packages
* `wheel` files are designed to install files only in the python `site-packages` (i.e. not in `<prefix>/share`)
* In the near future wheel files are going to be the only way to install files with pip
* `distutils` is deprecated. It will be removed from python 3.12 but still be available for while (how long ?) in the `setuptools` package

I managed to create custom `setuptools` commands to have `pip install`  installing our data files at the right location (`<prefix>/share`). But as `wheel` ignore those files `pip uninstall` ignore them too.

I'm sure that supporting old and new python packaging tooling at the same time going to be a nightmare.

@dreua Would it be possible to stop using wheel in the Fedora package ? It worked so far because we use `distutils` commands. If I switch to `setuptools` commands the `rpm` would no longer include translation. 

3 related goods news:
* A part from that `data_files` issue switching PDF Arranger to `pyproject.toml` does work on Linux (I checked it).
* `cx_Freeze` has started its migration to adapt to modern python packaging
* `python-distutils-extra` still relies on `distutils` but we'll soon no longer use it.

**Comment by dreua** [comment]
I'll take care of the Fedora package. Thanks for taking care of this, it sounds a bit like a mess and I haven't read the details but I trust you doing the right thing here.

**Reply by dreua** [reply]
Let's hope this doesn't affect too many other distributions. We should add a note to the next release to help package maintainers.

**Reply by dreua** [reply]
Another reminder that I need to take care of this:
```
/usr/lib/python3.12/site-packages/setuptools/_distutils/command/build.py:130: SetuptoolsDeprecationWarning: Direct usage of `distutils` commands
!!
        ********************************************************************************
        It seems that you are using `distutils.command.build` to add
        new subcommands. Using `distutils` directly is considered deprecated,
        please use `setuptools.command.build`.
        By 2023-Dec-13, you need to update your project and remove deprecated calls
        or your builds will no longer be supported.
        See https://peps.python.org/pep-0632/ for details.
        ********************************************************************************
!!
```
(Appeared when building the RPM.)

**Reply by dreua** [reply]
@jeromerobert an update about getting rid of wheel building in Fedora: When I initially packaged PDF Arranger for Fedora, building and installing wheels was possible but not the default way (iirc). I ended up doing that because that was how I got it to work and stuck with it. (Iirc that were similar issues with translations and data files just that wheels made it work, not the other way around.)

Therefore I first responded positively when you asked me to get rid of the wheels (here and in another issue I didn't find with a quick search).

Only later I reread the Fedora guidelines and and realized that nowadays building wheels is the recommended and preferred way and deviating from it should be well justified. Therefore I'd like to try sticking with wheels if possible. (We might even ask the Fedora Python-SIG for help if we can't figure it out ourselves.)

Actually I think with your latest changes it works even when following the latest guidelines and new RPM macros: #994 

Do you still think we need to get rid of wheels?

Sorry for the confusion!

**Reply by jeromerobert** [reply]
> Do you still think we need to get rid of wheels?

Not for the time being. Thanks to Cimbali I realize that the direction followed by Python packaging is still not clear to me. So let's wait and see.

> Sorry for the confusion!

Don't be sorry, here I'm the one causing confusion ;-).

**Comment by dreua** [comment]
Regarding uninstall:

Do I understand correctly that each pip install will leave some files behind that can't be uninstalled? That wouldn't sound ideal to me, at least we'd need to document this behavior IMO. Better would be fixing it. I'm worried this could lead to hard to track bugs in case we ever move, delete or rename files or in cases where you pip uninstall to switch to an rpm install for example. Left over files may hide others and/or become outdated.

**Reply by Cimbali** [reply]
I don’t think that’s right. Installed paths are recorded in the `{package}.dist-info/RECORD` file, so when uninstalling/updating they will get removed even if the `data_files` is not supported in the future.

**Reply by jeromerobert** [reply]
Hum, then I'll double recheck.

**Reply by jeromerobert** [reply]
Indeed it works with `pip install`. RECORD contains lines like:

```
../../../share/applications/com.github.jeromerobert.pdfarranger.desktop,sha256=
```

It works with `./setup.py bdist_wheel` on Debian. The wheel file constains things like:

```
pdfarranger-1.11a1.data/data/share/applications/com.github.jeromerobert.pdfarranger.desktop
```

It does not work with `python3 -m build` and according to our rpmbuild CI does not work with `./setup.py bdist_wheel` on Fedora. In both cases `*.mo` files are missing.

On Debia I can see that only `*.mo` files are missing. Other `data_files` are properly so I suspect a bug in our `setup.py`.

I wonder why https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-data-files says that `data_files` do not work with wheels.

**Reply by jeromerobert** [reply]
Well `python3 -m build --wheel` does work.

**Comment by dreua** [comment]
I've been using venvs recently and it seems to work quite well with PDF Arranger, might be recommendable for development and translation:

```
python3 -m venv --system-site-packages venv
source ./venv/bin/activate  # VS Code / Codium can automate this for you, just close and reopen the console
pip install .            # Repeat after code changes
pdfarranger          # works
man pdfarranger # also works
deactivate  # in case you want to leave the venv
```

Everything should be contained inside the venv. There is of course no Desktop integration (Nautilus "Open With", Application Menus etc.). For that it's probably the best to always build a system packge (RPM, DEB or whatever) and install it as such.

Unfortunately `pip install -e .` does not seem to work (fails to find a .ui file). You might be able to run the pdfarranger.py directly in the venv with installed helper files though, haven't tried that yet.

**Comment by Cimbali** [comment]
I think at the moment you can use setuptools + wheel with `data_files`, which from the discussion you’ve linked are more likely to be replaced rather than flat-out removed (at least I do that on another package I maintain I’m not sure what issues you run into, willing to look into them).

Also you could include more files inside the package directory to reduce the impact of the `data_files` removal (like the non OS-specific translations and ui files).

**Comment by jeromerobert** [comment]
> from the discussion you’ve linked are more likely to be replaced rather than flat-out removed

That is not my understanding. I understand that restricting the install directory and preventing PyPI packages from installing files anywhere they want in the `<prefix>` is a strong requirement. The reason for this is probably security.
 
> I do that on another package I maintain I’m not sure what issues you run into

Wow, indeed Pympress setup.py is really polished. I'll have a look in deep.

https://github.com/Cimbali/pympress/blob/master/setup.py#L279-L282 <-- I did not managed to make that work. For me setuptools `data_files` forbid installation out side of `site-packages`.

Yes we could move translations used directly in python and `.ui` files from `<prefix>/share` to `site-packages`. But freedesktop files and translations used directly by GTK (that is by `.ui` files) must remain in `<prefix>/share`. Also some distros automatically index translation files and expect to have them `<prefix>/share`. That would require packagers to patch.

**Reply by jeromerobert** [reply]
I'm probably wrong. I did all my tests using `pyproject.toml` instead of `setup.py`, may be it has more restriction.

**Comment by jeromerobert** [comment]
For the record, [Meld has dropped pip support](https://gitlab.gnome.org/GNOME/meld/-/issues/399).
